PCX-500: Update error handling (#115)
optile/android-sdk: README.rst

Payment Result
==============

> Payment results are returned through the onActivityResult() method in your Activity. When the payment page is closed, the returned PaymentResult class contains information about the operation request. I.e. it may contain an Interaction and OperationResult describing the state of a Charge operation. The PaymentError object inside the PaymentResult class is created when an error occurred inside the Android-SDK.

Reviewer: "PaymentResult class" looks wrong - do you mean "PaymentResult object"? The structure of the sentence looks a bit weird, the part "information about the operation request" is probably only understandable for optile insiders, and it remains unclear which subobjects of `PaymentResult` are optional and/or mandatory...

Idea for a better description: "When the payment page is closed, the `Intent` from its activity result can be converted to a `PaymentResult` containing (add list here, with clear distinction of what to expect in which case)"

Author: Rewrote the Payment Result chapter all together.

        }
        String resultInfo = result.getResultInfo();

        // Operation request has been made and "result" contains a resultInfo, an optional Interaction and OperationResult object.
        // The Interaction object is null when the user selected the same PresetAccount again or closed the page. 

>         if (resultCode == PaymentUI.RESULT_CODE_OK) {

Reviewer: Where does `resultCode` come from?

Reviewer: Sorry, I missed the paramter because I was a bit confused by `result` and `resultInfo`... maybe the code example could be made a bit more focused by introducing a `private void handlePaymentActivityResult(int resultCode, PaymentResult result)`?

Author: created a handlePaymentActivityResult method and removed the comments. 
It is better to explain which fields are provided in the README documentation than in this code sample.

            Interaction interaction = result.getInteraction();
            OperationResult operationResult = result.getOperationResult();
        } 

        // "result" contains a resultInfo and Interaction with optional OperationResult or PaymentError. 
        if (resultCode == PaymentUI.RESULT_CODE_CANCELED) {
            Interaction interaction = result.getInteraction();
            OperationResult operationResult = result.getOperationResult();
            PaymentError error = result.getPaymentError();
        }
    }

Successful
----------

> The RESULT_CODE_OK code indicates that the operation request was successful, there are three situations when this result is returned:

Reviewer: Plus/minus one problem? :trollface:

Author: Removed these texts all together

1. The user closed the page or selected an already selected PresetAccount from the list, in both cases the Interaction and OperationResult objects are null.

2. InteractionCode is PROCEED. The PaymentResult contains an OperationResult with detailed information about the successful operation.

Canceled
---------

The RESULT_CODE_CANCELED code indicates that the Android SDK did not perform a successful operation. The PaymentResult will always contain an Interaction and optional OperationResult or PaymentError indicating what went wrong and what should be done next. 

> 1. When the code of the Interaction is VERIFY, it means that the Android-SDK was not able to determine the current state of the list and should be verified before continuing.

Reviewer: Something's missing here... what exactly should be verified?

Author: Rewrote the payment result chapter

> 1. When the code of the Interaction is VERIFY, it means that the Android-SDK was not able to determine the current state of the list and should be verified before continuing.

Reviewer: More generally about this list: can we define more clearly how the three reasons would be determined? The conditions you give here don't look like they cleanly separate between those three reasons.

Author: Rewrote the PaymentResult chapter

> 2. The user performed an operation and the result contains an OperationResult containing detailed information why the operation failed.

Reviewer: Maybe two sentences?
```suggestion
2. The user performed an operation that failed. The result contains an OperationResult with detailed information about the cause of the failure.
```

Author: Rewrote the Payment Result chapter

> 3. An error occurred inside the Android-SDK and the result contains a PaymentError. When the PaymentResult contains a PaymentError, the PaymentResult will also contain a client-side created Interaction object. The following paragraph describes which types of Interactions are created by the Android-SDK when an internal error occurred.

Reviewer: Same here - I would prefer a very concise first sentence:
```suggestion
3. An error occurred inside the Android-SDK. In this case the PaymentResult contains a PaymentError and a client-side created Interaction object. The following paragraph describes which types of Interactions are created by the Android-SDK when an internal error occurred.
```

Author: Rewrote the payment result chapter

    
Client-Side Interactions 
------------------------

The following table describes the combination of InteractionCode and InteractionReason created by the Android-SDK.

+------------------+-----------------------+-----------------------------------------------------------------+
| InteractionCode  | InteractionReason     | Description                                                     |
+==================+=======================+=================================================================+
| ABORT            | CLIENTSIDE_ERROR      | An internal error occurred inside the Android-SDK, i.e. a       |
|                  |                       | SecurityException was thrown. The list may still be valid.      |      
+------------------+-----------------------+-----------------------------------------------------------------+
| ABORT            | COMMUNICATION_FAILURE | A network failure occurred while communicating with the         |            
|                  |                       | Optile Payment API. The list may still be valid.                |
+------------------+-----------------------+-----------------------------------------------------------------+

> | VERIFY           | CLIENTSIDE_ERROR      | An error occurred during a Charge operation.                    |

Reviewer: So "VERIFY" belongs to case 3 (interaction created on the client-side) and also to case 1 (VERIFY) from the list above? Hmm... something does not add up here.

|                  |                       | The charge may have been successful, therefor the status of the | 

> |                  |                       | payment (list) must be verified.                           |

Reviewer: Minor: missing spaces before the final character

Author: This is fixed
